src/util/indexing.jl

```diff
@@ -0,0 +1,249 @@
+                names[i] = newnm
+                push!(seen, newnm)
+                break
+            end
+            k += 1
+        end
+    end
+    return names
+end
+
+
+Base.length(x::Indexer) = length(x.names)
+Base.names(x::Indexer) = copy(x.names)
+_names(x::Indexer) = x.names
+Base.copy(x::Indexer) = Indexer(copy(x.lookup), copy(x.names))
+Base.deepcopy(x::Indexer) = copy(x) # all eltypes immutable
+Base.isequal(x::Indexer, y::Indexer) = isequal(x.lookup, y.lookup) && isequal(x.names, y.names)
+Base.(:(==))(x::Indexer, y::Indexer) = isequal(x, y)
+Base.haskey(x::Indexer, key::Symbol) = haskey(x.lookup, key)
+Base.haskey{S <: AbstractString}(x::Indexer, key::S) = haskey(x, convert(Symbol, key))
+Base.haskey(x::Indexer, key::Real) = 1 <= key <= length(x.names)
+Base.keys(x::Indexer) = names(x)
+
+
+"Completely replace the names in an Indexer with a new Vector of Symbols."
+function names!(x::Indexer, names::Vector{Symbol})
+    if length(names) != length(x)
+        throw(ArgumentError("Length of new names doesn't match length of Index."))
+    end
+    names = make_unique(names)
+    if length(x.names) != length(x.lookup)
+        throw(ArgumentError("Index corrupted by duplicate symbols in names."))
+    end
+    rename!(x, x.names, names)
+    return x
+end
+
+
+"Completely replace the names in an Indexer with a new Vector of Strings."
+function names!{S <: AbstractString}(x::Indexer, names::Vector{S})
+    names!(x, convert(Vector{Symbol}, names))
+end
+
+
+function rename!(x::Indexer, names)
+    for (from, to) in names
+        if !haskey(x, from)
+            throw(ArgumentError("Cannot rename $from to $to, $from does not exist in the Indexer."))
+        end
+        if haskey(x, to)
+            throw(ArgumentError("Cannot rename $from to $to, $to already exists in the Indexer."))
+        end
+        # Change the name array
+        nameidx = findfirst(x.names, from)
+        x.names[nameidx] = to
+        # Change the dictionary
+        x.lookup[to] = val = pop!(x.lookup, from)
+    end
+    return x
+end
+
+
+"""
+Rename a set of names in an Indexer, to a new set of names.
+
+This will rename such that the name in from[1] will be renamed to to[1],
+from[2] will be renamed to to[2] and so on.
+"""
+function rename!(x::Indexer, from::Vector{Symbol}, to::Vector{Symbol})
+    return rename!(x, zip(from, to))
+end
+
+
+"Rename a single name (from) in Indexer x, to a new name (to)."
+rename!(x::Indexer, from::Symbol, to::Symbol) = rename!(x, ((from, to),))
+
+
+"""
+Rename a indexer according to some function, f, that accepts one argument.
+
+Each name in the Indexer x, will be input to, f, and the output will be
+what that name will be renamed to.
+
+Therefore the function, f, will need to return Symbols or Strings.
+"""
+rename!(x::Indexer, f::Function) = rename!(x, [(x,f(x)) for x in x.names])
+rename!(f::Function, x::Indexer) = rename!(x, f)
```

Comment:
It's basically if people put the two arguments the other way round. So the history of this type is I took it verbatim from - I think - DataFrames.jl, and modified it a bit, and then recently modified it a lot, and so this second rename! is a revenant of when it was more like DataFrames.jl. I don't know which is more julian, but since the second function is the odd one out of all the code, we can get rid of that one.